example/src/DeviceUpdates.hs

```diff
@@ -0,0 +1,230 @@
+
+initModel :: Model ref
+initModel = Model M.empty M.empty M.empty
+
+preconditions
+  :: forall refs resp. IxForallF Ord refs
+  => Model refs -> Action refs resp -> Bool
+preconditions m cmd = (case cmd of
+  AddDevice                -> True
+  AddUpdate _              -> True
+  ScheduleUpdate uref dref ->
+    uref `elem` M.keys (updates m) &&
+    dref `elem` M.keys (installedUpdates m)
+  UpdateReports dref       -> dref `elem` M.keys (installedUpdates m)
+  PullUpdates dref         -> dref `elem` M.keys (installedUpdates m)
+  ViewUpdates dref         -> dref `elem` M.keys (installedUpdates m)
+  ) \\ (iinstF @'DeviceRef Proxy :: Ords' refs 'DeviceRef)
+    \\ (iinstF @'UpdateRef Proxy :: Ords' refs 'UpdateRef)
+
+transitions
+  :: forall refs resp. IxForallF Ord refs
+  => Model refs -> Action refs resp -> Response_ refs resp -> Model refs
+transitions m cmd resp = (case cmd of
+
+  AddDevice                ->
+    m { installedUpdates = M.insert resp S.empty (installedUpdates m)
+      , pendingUpdates   = M.insert resp S.empty (pendingUpdates   m)
+      }
+
+  AddUpdate update         ->
+    m { updates = M.insert resp update (updates m) }
+
+  ScheduleUpdate uref dref ->
+    let update = S.singleton (updates m M.! uref) in
+    m { pendingUpdates   = M.insertWith S.union dref update (pendingUpdates m) }
+
+  UpdateReports _          -> m
+
+  PullUpdates dref         ->
+    m { installedUpdates = M.insertWith S.union dref (pendingUpdates m M.! dref)
+                                                     (installedUpdates m)
+      , pendingUpdates   = M.insert dref S.empty (pendingUpdates m)
+      }
+
+  ViewUpdates _            -> m
+
+  ) \\ (iinstF @'DeviceRef Proxy :: Ords' refs 'DeviceRef)
+    \\ (iinstF @'UpdateRef Proxy :: Ords' refs 'UpdateRef)
+
+postconditions
+  :: forall refs resp. IxForallF Ord refs
+  => Model refs -> Action refs resp -> Response_ refs resp -> Property
+postconditions m cmd resp = (case cmd of
+  ViewUpdates dref -> resp === installedUpdates m M.! dref
+  PullUpdates dref -> pendingUpdates m' M.! dref === S.empty
+  _                -> property True
+  ) \\ (iinstF @'DeviceRef Proxy :: Ords' refs 'DeviceRef)
+  where
+  m' = transitions m cmd resp
+
+smm :: StateMachineModel Model Action
+smm = StateMachineModel preconditions postconditions transitions initModel
+
+------------------------------------------------------------------------
+
+gen :: Gen (Untyped Action (RefPlaceholder Refs))
+gen = frequency
+  [ (1, return . Untyped $ AddDevice)
+  , (5, Untyped . AddUpdate <$> arbitrary)
+  , (5, return . Untyped $ ScheduleUpdate SUpdateRef SDeviceRef)
+  , (5, return . Untyped $ UpdateReports SDeviceRef)
+  , (5, return . Untyped $ PullUpdates SDeviceRef)
+  , (5, return . Untyped $ ViewUpdates SDeviceRef)
+  ]
+
+shrink1 :: Action refs resp -> [Action refs resp]
+shrink1 _ = []
+
+------------------------------------------------------------------------
+
+semantics
```

Comment:
Maybe another name for this to signal that this is not the "real" semantics, `debugSemantics` or something like that.